src/PerfTreeParser/PerfTreeParser.class.st

```diff
@@ -0,0 +1,244 @@
+			^ everyChildren ].
+
+	aNode to: lastNode do: [ :i |
+		(self numberOfSpacesAt: i) - (self numberOfSpacesAt: aNode) = 11
+			ifTrue: [ everyChildren add: i ] ].
+
+	^ everyChildren
+]
+
+{ #category : #accessing }
+PerfTreeParser >> findChildrenOfIndex: aNodeIndex [
+	"Gives the nodes representing every children at the node aNodeIndex."
+
+	| collectionOfChildrenIndexes |
+	collectionOfChildrenIndexes := self findChildrenIndexOfIndex: aNodeIndex.
+	^ collectionOfChildrenIndexes collect: [ :childIndex |
+		  self nodeAtIndex: childIndex ]
+]
+
+{ #category : #accessing }
+PerfTreeParser >> lastSpaceAt: aLine [
+	"Gives the index of the last space or '-' at a given line."
+
+	| lastSpace line |
+	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	line := self readLine: aLine.
+	lastSpace := 0.
+
+	1 to: line size do: [ :i |
+		((line at: i) = Character space or: [ (line at: i) = $- ]) ifTrue: [
+			lastSpace := i ] ].
+	^ lastSpace
+]
+
+{ #category : #accessing }
+PerfTreeParser >> nameOf: aNodeIndex [
+	"Gives the name of a function at aNodeIndex."
+
+	| aLine |
+	(aNodeIndex between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	aLine := self readLine: aNodeIndex.
+	^ aLine last: aLine size - (self lastSpaceAt: aNodeIndex)
+]
+
+{ #category : #accessing }
+PerfTreeParser >> nodeAtIndex: aNodeIndex [
+	"Gives the node representing the node at aNodeIndex."
+
+	| child |
+	(aNodeIndex between: 1 and: self numberOfFunctions) ifFalse: [ ^ nil ].
+
+	child := PerfTreeNode new
+		         name: (self nameOf: aNodeIndex);
+		         weight: (self weightOf: aNodeIndex);
+		         time: self time;
+		         yourself.
+	child children: (self findChildrenOfIndex: aNodeIndex).
+	child children do: [ :grandchild | grandchild parent: child ].
+	^ child
+]
+
+{ #category : #accessing }
+PerfTreeParser >> numberOfFunctions [
+	"Gives the number of functions present in the file."
+
+	^ content size
+]
+
+{ #category : #accessing }
+PerfTreeParser >> numberOfSpacesAt: aNodeIndex [
+	"Gives the number of spaces at a certain line."
+
+	| count lineRead |
+	count := 0.
+
+	(aNodeIndex between: 1 and: self numberOfFunctions) ifFalse: [
+		^ count ].
+
+	lineRead := self readLine: aNodeIndex.
+	count := 1.
+
+	[(lineRead at: count) = Character space or: [
+		(lineRead at: count) = $| ] ] whileTrue: [ count := count + 1 ].
+	^ count - 1
+]
+
+{ #category : #accessing }
+PerfTreeParser >> percentageOf: aLine [
+	"Gives the percentage from a function at a given line"
+
+	"In the case of multiples functions for one percentage, it gives the same percentage to every functions."
+
+	| stringLine numbers |
+	(aLine between: 1 and: self numberOfFunctions) ifFalse: [ ^ 0 ].
+
+	stringLine := ((self content at: aLine) copyUpTo: $%) last: 5.
+
+	stringLine = ((self content at: aLine) last: 5) ifTrue: [
+		^ self percentageOf: aLine - 1 ].
+
+	numbers := stringLine asNumber.
+	numbers < 0 ifTrue: [ numbers := numbers negated ].
+
+	^ numbers
```

Comment:
Here should be `number` (in singular). And why are you negating the negative? What means a negative number here?